corehq/apps/app_manager/app_schemas/case_properties.py

```diff
@@ -127,26 +127,39 @@ def _all_possible_equivalences(self):
 
         """
         all_possible_equivalences = set()
-        equivalence_queue = deque(_CaseTypeEquivalence(case_type, _CaseTypeRef(case_type, ()))
-                                  for case_type in self.parent_type_map)
+
+        class TmpEquivalence(namedtuple('TmpEquivalence',
+                                        ['case_type', 'expansion', 'visited_case_types'])):
+            """same as a _CaseTypeEquivalence but with a memory of the case types visited"""
+
+        equivalence_queue = deque(
+            TmpEquivalence(case_type, _CaseTypeRef(case_type, ()), visited_case_types=())
+            for case_type in self.parent_type_map
+        )
+
         while True:
             try:
-                equivalence = equivalence_queue.popleft()
+                eq = equivalence_queue.popleft()
             except IndexError:
                 break
-            all_possible_equivalences.add(equivalence)
-            parent, (child, relationship_path) = equivalence
-            for relationship, grandparents in self.parent_type_map[parent].items():
-                for grandparent in grandparents:
-                    new_equivalence = _CaseTypeEquivalence(
-                        case_type=grandparent,
-                        expansion=_CaseTypeRef(child, relationship_path + (relationship,))
+            all_possible_equivalences.add(eq)
+            for relationship, parents in self.parent_type_map[eq.case_type].items():
+                for parent in parents:
+                    if parent in eq.visited_case_types:
+                        continue
+                    new_equivalence = TmpEquivalence(
+                        case_type=parent,
+                        expansion=_CaseTypeRef(
+                            eq.expansion.case_type,
+                            eq.expansion.relationship_path + (relationship,)
+                        ),
+                        visited_case_types=tuple(sorted(set(eq.visited_case_types + (parent,))))
```

Comment:
is there a reason this needs to be sorted or a tuple as opposed to just storing the set?

Reply:
Yeah, annoyingly you can't have a `set` of `set`s because `set`s are mutable and thus not hashable, so they can't be used as `dict` keys or `set` elements. The sorted tuple with no repeat elements here was meant as a replacement for that. That said, I think `frozenset` would be as good or better in every way, so I'll replace this with that.